### DISCUSSION TITLE
Proposal/discussion: Schemas: JSON Schema and maybe more...

### DISCUSSION THREAD

**Discussion by liamg** [open-policy-agent/community#278] [OPA and Rego]
_I've been discussing this with @srenatus and we thought it best to open it up for global discussion..._

Currently the JSON Schema parsing in OPA is done by [forking/internalising][1] the [xeipuuv/gojsonschema][2] module. There is then a conversion function in `opa` which takes a data structure from the module and spits out a [`types.Type`][7] for validation use.

This module appears to no longer be maintained, and has many open issues. It also doesn't support drafts [2020-12][3] and [2019-09][4].

Recently there have been a couple of missing features/bugs (see [link][5], [link][6]) in JSON Schema functionality that have revealed the above to be a potential issue.

The required functionality in OPA, currently can be split into two areas:

1. Parse a JSON Schema into a meaningful data structure
  a. ...with the caveat that we must be able to avoid network access during parsing as per the specified capabilities
1. Convert the parsed schema into a [`types.Type`][7] to be used for policy validation.

My proposal is:

1. Switch to a better-maintained library for area (1) - such as [santhosh-tekuri/jsonschema][8] which also provides API to cater for (1a) above.
2. Write a default schema parsing function to handle converting the [santhosh-tekuri/jsonschema][8] data structure to a [`types.Type`][7]- i.e. replacing [`parseSchema`][9]
3. Allow a consumer to customise behaviour by overriding the internal schema parsing function (therefore handling (1), (1a) and (2)), perhaps by implementing the below interface:
```go
type SchemaParser interface {
  Parse(schema interface{}, allowNet bool) (*types.Type, error)
}
```
...and calling `compiler.WithSchemaParser(myCustomParser)`. This would also have the advantage of allowing for other schema/validation formats in other consuming projects if required, without OPA having to know about them in advance.

WDYT?

[1]: https://github.com/open-policy-agent/opa/tree/main/internal/gojsonschema
[2]: https://github.com/xeipuuv/gojsonschema
[3]: https://json-schema.org/specification-links.html#2020-12
[4]: https://json-schema.org/specification-links.html#draft-2019-09-formerly-known-as-draft-8
[5]: https://github.com/open-policy-agent/opa/pull/5188
[6]: https://github.com/open-policy-agent/opa/pull/5167
[7]: https://github.com/open-policy-agent/opa/blob/b2d92a33c14e5b5320250b8ef32fe5fe0026c628/types/types.go#L27-L31
[8]: https://github.com/santhosh-tekuri/jsonschema
[9]: https://github.com/open-policy-agent/opa/blob/b2d92a33c14e5b5320250b8ef32fe5fe0026c628/ast/compile.go#L1056

**Comment by anderseknert** [comment]
Sounds good to me! Do you see any risks with the approach?
